internal/gouging/gouging.go

```diff
@@ -0,0 +1,124 @@
+package gouging
+
+import (
+	"fmt"
+	"math"
+
+	"gitlab.com/NebulousLabs/errors"
+	rhpv2 "go.sia.tech/renterd/rhp/v2"
+	"go.sia.tech/siad/modules"
+	"go.sia.tech/siad/types"
+)
+
+var (
+	// ErrGougingDetected is returned when price gouging was detected in one of
+	// the fields of either the price table or the host's external settings.
+	ErrGougingDetected = errors.New("price gouging detected")
+
+	downloadOverdriveEstimatePct = .1
+)
+
+type (
+	GougingChecks struct {
+		Download     GougingCheck
+		FormContract GougingCheck
+		Upload       GougingCheck
+	}
+
+	GougingCheck interface {
+		IsGouging() bool
+		error
+	}
+
+	GougingSettings struct {
+		MaxRPCPrice      types.Currency
+		MaxContractPrice types.Currency
+		MaxDownloadPrice types.Currency // per TiB
+		MaxUploadPrice   types.Currency // per TiB
+	}
+
+	check struct {
+		err error
+	}
+)
+
+func (c *check) IsGouging() bool {
+	return c.err != nil
+}
+
+func (c *check) Error() string {
+	if c.err != nil {
+		return c.err.Error()
+	}
+	return ""
+}
+
+func PerformGougingChecks(gs GougingSettings, hs rhpv2.HostSettings, period uint64, redundancy float64) GougingChecks {
+	return GougingChecks{
+		Download:     &check{err: checkDownloadGouging(gs, hs, redundancy)},
+		FormContract: &check{err: checkFormContractGouging(gs, hs)},
+		Upload:       &check{err: checkUploadGouging(gs, hs, period, redundancy)},
+	}
+}
+
+func checkDownloadGouging(gs GougingSettings, hs rhpv2.HostSettings, redundancy float64) error {
+	// check base rpc price
+	if !gs.MaxRPCPrice.IsZero() && hs.BaseRPCPrice.Cmp(gs.MaxRPCPrice) > 0 {
+		return fmt.Errorf("rpc price exceeds max: %v>%v, %w", hs.BaseRPCPrice, gs.MaxRPCPrice, ErrGougingDetected)
+	}
+
+	// check download cost
+	minRedundancy := uint64(math.Ceil(redundancy))
```

Comment:
Why not `MulFloat` for more accuracy? If the redundancy is something like `3.5` this is going to be quite inaccurate and lower the computed price by more than 10%.

Reply:
Hm not sure why I didn't go for `MulFloat` - I probably had a bad reason for it while writing it and then due to tiny refactors it became moot. Will update, it wouldn't be lower though right, it was always intended to be a worst case. I added the overdrive % because originally I was multiplying with `MinShards` here